src/garage/sampler/ray_batched_sampler.py

```diff
@@ -0,0 +1,216 @@
+        self._algo = algo
+        self._seed = seed
+        deterministic.set_seed(self._seed)
+        self._max_path_length = self._algo.max_path_length
+        self._should_render = should_render
+        if not ray.is_initialized():
+            ray.init(ignore_reinit_error=True)
+        self._num_workers = num_processors if num_processors \
+            else psutil.cpu_count(logical=False)
+        self._all_workers = defaultdict(None)
+
+    def start_worker(self):
+        """Initialize a new ray worker."""
+        env_pkl_id = ray.put(pickle.dumps(self._env))
+        agent_pkl_id = ray.put(pickle.dumps(self._algo.policy))
+        for worker_id in range(self._num_workers):
+            self._all_workers[worker_id] = (self._sampler_worker.remote(
+                worker_id, env_pkl_id, agent_pkl_id, self._seed,
+                self._max_path_length, self._should_render))
+        self._idle_worker_ids = list(range(self._num_workers))
+
+    def obtain_samples(self, itr, num_samples):
+        """Sample the policy for new trajectories.
+
+        Args:
+            - itr(int): iteration number
+            - num_samples(int):number of steps the the sampler should collect
+        """
+        self._active_workers = []
+        self._active_worker_ids = []
+        pbar = ProgBarCounter(num_samples)
+        completed_samples = 0
+        traj = []
+        updating_workers = []
+        self._idle_worker_ids = list(range(self._num_workers))
+        curr_policy_params = self._algo.policy.get_param_values()
+        params_id = ray.put(curr_policy_params)
+        while self._idle_worker_ids:
+            worker_id = self._idle_worker_ids.pop()
+            worker = self._all_workers[worker_id]
+            updating_workers.append(worker.set_agent.remote(params_id))
+        while completed_samples < num_samples:
+            updated, updating_workers = ray.wait(
+                updating_workers, num_returns=1, timeout=0.1)
+            upd = [ray.get(up) for up in updated]
+            self._idle_worker_ids.extend(upd)
+            while self._idle_worker_ids:
+                idle_worker_id = self._idle_worker_ids.pop()
+                self._active_worker_ids.append(idle_worker_id)
+                worker = self._all_workers[idle_worker_id]
+                self._active_workers.append(worker.rollout.remote())
+            ready, not_ready = ray.wait(
+                self._active_workers, num_returns=1, timeout=0.001)
+            self._active_workers = not_ready
+            for result in ready:
+                trajectory, num_returned_samples = self._process_trajectory(
+                    result)
+                completed_samples += num_returned_samples
+                pbar.inc(num_returned_samples)
+                traj.append(trajectory)
+        pbar.stop()
+        return traj
+
+    def shutdown_worker(self):
+        """Shuts down the worker."""
+        ray.shutdown()
+
+    def _process_trajectory(self, result):
```

Comment:
can you add a docstring about what this is doing? honestly i'm not sure why it's necessary.

Reply:
It was so I could pass pep8. I had too many nested loops, and also, it then made sense to have a function that processes trajectories returned from workers, after they are deposited in the object store.

Reply:
resolved